File: dsa/concepts/06_divandcon/java/Marge_L493_ReversePairs.java
import java.util.ArrayList;

public class Marge_L493_ReversePairs {

	public static void main(String[] args) {
		Marge_L493_ReversePairs obj = new Marge_L493_ReversePairs();

		int[] nums = new int[] {1,3,2,3,1};
		System.out.println(obj.reversePairs(nums));
	}

	// Problem: https://leetcode.com/problems/reverse-pairs/
	// Idea: Use merge sort approach. Two sorted arrays are used to calculate pairs. For every element in arr1, check matching pairs in arr2.
	// Pairs from arr1 and arr2 are calculated recursively. Pairs for arr1 and arr2 combined is calculated before merge process.
	public int reversePairs(int[] nums) {
        return mergeSort(nums, 0, nums.length-1);
    }

    private int mergeSort(int[] nums, int low, int high) {
		if(low >= high){
			return 0;
		}
		int mid = (low+high)/2;
		int pairs = mergeSort(nums, low, mid);
		pairs += mergeSort(nums, mid+1, high);
		pairs += merge(nums, low, mid, high);

        return pairs;
	}

	private int merge(int nums[], int low, int mid, int high){
		// Calculate pairs
		int pairs = 0;
        for(int i=low, j=mid+1; i<=mid; i++){
			while(j<=high && nums[i] > (long)2*nums[j])
				j++;

            pairs += j-(mid+1);
        }

		// Merge arrays
		int left = low, right = mid+1;
		ArrayList<Integer> temp = new ArrayList<>();
		while(left <= mid && right <= high){
			if(nums[left] <= nums[right]){
				temp.add(nums[left++]);
			}else{
				temp.add(nums[right++]);
			}
		}
		while(left <= mid){
			temp.add(nums[left++]);
		}
		while(right <= high){
			temp.add(nums[right++]);
		}

		for(int i=low; i<=high; i++) {
			nums[i] = temp.get(i-low);
		}

        return pairs;
	}
}

File: dsa/concepts/06_divandcon/python/marge_l493_reverse_pairs.py
"""
Problem: https://leetcode.com/problems/reverse-pairs/
Idea:
Use merge sort.
While merging two sorted halves, count pairs such that
nums[i] > 2 * nums[j] where i is in the left half and j is in the right half.
Pairs inside left and right halves are counted recursively.
Cross pairs are counted before merging.
"""

def reverse_pairs(nums):
    def merge_sort(left, right):
        if left >= right:
            return 0

        mid = (left + right) // 2
        pairs = merge_sort(left, mid)
        pairs += merge_sort(mid + 1, right)
        pairs += count_and_merge(left, mid, right)
        return pairs

    def count_and_merge(left, mid, right):
        # Count reverse pairs
        count = 0
        j = mid + 1

        for i in range(left, mid + 1):
            while j <= right and nums[i] > 2 * nums[j]:
                j += 1
            count += j - (mid + 1)

        # Merge the two sorted halves
        temp = []
        l, r = left, mid + 1

        while l <= mid and r <= right:
            if nums[l] <= nums[r]:
                temp.append(nums[l])
                l += 1
            else:
                temp.append(nums[r])
                r += 1

        temp.extend(nums[l:mid + 1])
        temp.extend(nums[r:right + 1])

        nums[left:right + 1] = temp
        return count

    return merge_sort(0, len(nums) - 1)


# Example usage
if __name__ == "__main__":
    nums = [1, 3, 2, 3, 1]
    print(reverse_pairs(nums))  # 2
